ClusterMesh/KVStoreMesh Readiness and Bootstrap QPS

thorn3r opened the pull request:
This adds a readiness probe to the kvstoremesh container which will report ready once all remote clusters have been synchronized. A per-cluster timeout on initial sync prevents a deadlock in a situation where the pod is restarted simultaneously on multiple clusters. In addition to the per-cluster timeout, a global timeout is also implemented to catch any other failures. 

To increase the speed of the initial synchronization for both clustermesh-apiserver and kvstoremesh, a new kvstore-opt was added, `etcd.bootstrapQps`, which allows an initial rate-limit to be set for the etcd client. Once the container is ready, the standard rate-limit configured via `etcd.qps` is applied. `etcd.bootstrapQps` is set to 10,000 for both containers in the chart.

Some rough benchmarking with simulated load provided promising results:
- This test was run with 2 connected clusters
- Load on clustermesh-apiserver/kvstoremesh was simulated by generating 10,000 CiliumIdentities to be synchronized. cilium-operator was disabled to prevent garbage collection

For clustermesh-apiserver, `time-to-ready` dropped from ~500s to ~14s (roughly a 60x improvement)
![apiserver-10000identities](https://github.com/cilium/cilium/assets/22533018/8e4f96eb-e9a8-4fcb-90c6-dbcbd04922cf)


For kvstoremesh, `time-to-ready` dropped from ~100s to 13s (roughly a 10x improvement)
![kvstoremesh-10000identities](https://github.com/cilium/cilium/assets/22533018/953f46f3-a358-4297-9130-8cb100b51a20)
 


Additional details are in each commit message.

```release-note
Add a readinessProbe to the kvstoremesh container that reports initial synchronization status to support configuring a separate, initial rate-limit to be used while synchronizing. Both clustermesh-apiserver and kvstoremesh now use a high initial rate-limit to decrease start time.
```

thorn3r committed:
remove duplicate import from clustermesh/common

Remove duplicate import of clustermesh/types from clustermesh/common

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

squeed commented on clustermesh-apiserver/health/health.go:
```diff
@@ -10,54 +10,53 @@ import (
 		if syncState.Complete() {
 			statusCode = http.StatusOK
 			reply = "Ready"
+		} else {
```
empty else?

thorn3r replied:
whoops, overlooked leftover from testing :sweat_smile:

squeed commented on install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:
```diff
@@ -208,7 +208,7 @@ spec:
           httpGet:
             path: /readyz
-            port: apiserv-health
+            port: 9881
```
should this be a templated helm value? (Yes it should :-) )

thorn3r replied:
This shouldn't have been included in this commit. The following overwrites this, i'll fix it for history

squeed commented on pkg/clustermesh/common/metrics.go:
```diff
@@ -52,3 +54,22 @@ func MetricsProvider(subsystem string) func() Metrics {
+		return ApiserverMetrics{
+			InitialSyncDuration: metric.NewGauge(metric.GaugeOpts{
+				Namespace: metrics.Namespace,
+				Name:      "initial_sync_time_seconds",
```
This should probably be prefixed with `clustermesh_`.

thorn3r replied:
This metric is exported for both `clustermesh-apiserver` and `kvstoremesh`, so we're utilizing the Namespace which is set on startup [here](https://github.com/cilium/cilium/blob/0ea1bc9c6e73121c3adbdcc07cc5a38a633c853b/clustermesh-apiserver/clustermesh/root.go#L59) for clustemesh-apiserver and [here](https://github.com/cilium/cilium/blob/0ea1bc9c6e73121c3adbdcc07cc5a38a633c853b/clustermesh-apiserver/kvstoremesh/root.go#L33) for kvstoremesh

The exported metrics look like this:
```
cilium_clustermesh_apiserver_initial_sync_time_seconds
cilium_kvstoremesh_initial_sync_time_seconds
```

squeed replied:
Aha! Never mind then :)

squeed commented on pkg/clustermesh/common/metrics.go:
```diff
@@ -52,3 +54,22 @@ func MetricsProvider(subsystem string) func() Metrics {
+			InitialSyncDuration: metric.NewGauge(metric.GaugeOpts{
+				Namespace: metrics.Namespace,
+				Name:      "initial_sync_time_seconds",
+				Help:      "Duration in seconds of kvstore initial sync",
```
`kvstore` here is ambiguous; can you add a bit more context -- and reference that this is for kvstoremesh + clustermesh.

thorn3r replied:
The synchronization is different depending on which process it is: for clustermesh-apiserver, we're syncing resources from local cluster to the kvstore. For kvstoremesh, we're syncing the resources from remote clusters to the local kvstore. I'll try to add some additional details here without being too verbose

squeed commented on pkg/clustermesh/common/sync.go:
```diff
@@ -0,0 +1,99 @@
+
+// ClusterSyncState is used to track the synchronization of a single cluster to
+// the local kvstore.
+type ClusterSyncState struct {
```
Does this type need to be exported?

thorn3r replied:
Nope, just its methods. I'll update that :+1:

squeed commented on install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:
```diff
@@ -291,8 +299,11 @@ spec:
         {{- with .Values.clustermesh.apiserver.kvstoremesh.extraEnv }}
         {{- toYaml . | trim | nindent 8 }}
         {{- end }}
-        {{- if .Values.clustermesh.apiserver.metrics.kvstoremesh.enabled }}
```
is this `if` now in the wrong place?

thorn3r replied:
I don't think so. The `kvmesh-health` port is used for readiness checks and should be included regardless of whether metrics are enabled or not. This conditional is now just for adding the metrics port to the list.

squeed commented on pkg/kvstore/etcd.go:
```diff
@@ -371,7 +371,10 @@ type etcdClient struct {
-	limiter       *ciliumrate.APILimiter
+	// limiterLock protects limiter from concurrent access
+	limiterLock lock.RWMutex
+	limiter     *ciliumrate.APILimiter
```
what about an `atomic.Pointer[ciliumrate.APILimiter]`? Might be easier to reason about.

thorn3r replied:
I like it :+1:

marseel commented on pkg/clustermesh/common/remote_cluster.go:
```diff
@@ -132,6 +135,19 @@ func (rc *remoteCluster) releaseOldConnection() {
+	// If the timeout has been reached and a connection hasn't been established, we
+	// remove the cluster from the SyncState to prevent it from blocking readiness forever.
+	go func() {
+		<-time.After(option.Config.KVstoreConnectivityTimeout)
```
hmm, I'm unsure if we shouldn't have a separate parameter for it like `KVStoreMeshConnectivityTimeout`. Resuing KVstore one might be confusing.

marseel replied:
I would probably also explicitly add it to helm as a parameter - once we have HA, users might want to increase it also depending on clustermesh size.

thorn3r replied:
Yeah I was also back and forth on this, since the option does seem applicable but is also used in a different context elsewhere. I'll make a new option for this

marseel commented on pkg/clustermesh/common/remote_cluster.go:
```diff
@@ -132,6 +135,19 @@ func (rc *remoteCluster) releaseOldConnection() {
+		<-time.After(option.Config.KVstoreConnectivityTimeout)
+		rc.mutex.Lock()
+		defer rc.mutex.Unlock()
+		if !connectionEstablished {
```
Probably a bit more straightforward would be to have `connectionEstablished` as channel and just wait on both timeout and receiving information on channel about connectionEstablished - we will also exit early this goroutine if connections will be established + no need for locking

thorn3r replied:
Ah yes you're right. I initially was trying to catch scenarios where we establish connection but fail afterward. But then I realized that this will only happen if we fail to write the cluster config to the kvstore, in which case we probably _do_ want to block readiness. I'll simplify this

giorio94 replied:
> But then I realized that this will only happen if we fail to write the cluster config to the kvstore, in which case we probably do want to block readiness.

There's also the case where we fail to retrieve the cluster config, which should be probably treated as failing to connect to etcd in terms of not blocking readiness (it typically happens when there's something wrong in the cluster name configuration).

thorn3r replied:
What do you think about using a channel for `connectionEstablished` and closing is at the end of the DoFunc before successful exit? There's a chance we might hit the timeout and un-block a cluster that's actually going to synchronize (i.e. writing the cluster config to the kvstore and we're waiting for the ready channel to be closed), but this seems like an edge-case and a pretty benign one - and will make the logic much simpler to follow

giorio94 replied:
How about https://github.com/cilium/cilium/pull/30361#issuecomment-1912286336? It should be even simpler as it doesn't require to modify anything in the common package.

marseel commented on install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:
```diff
@@ -263,19 +263,28 @@ spec:
+          httpGet:
+            path: /readyz
+            port: kvmesh-health
+          periodSeconds: 30
```
We can probably reduce it or was there a specific reason for picking 30s?

thorn3r replied:
Tried to pick a useful but non-aggressive period here, since this will continue to run indefinitely even after the container is ready. Do you think a lower value would be better?

marseel replied:
ah good point, how about we use startupProbe instead?

thorn3r replied:
I considered using a startupProbe (for clustermesh-apiserver readiness as well), but it will restart the Pod on failure (like livenessProbe) instead of just keeping it out of rotation until its ready. It would be nice if there was a `startup` level probe for readiness as well. It would also be difficult to pick a value that would be appropriate for small and large scale clustermesh deployments

giorio94 replied:
Could be worth adding a startup probe now that we have a global circuit breaker, so that we can check more aggressively whether kvstoremesh turned ready.

giorio94 replied:
Could it also make sense to make these values configurable? For instance, we may want to set lower values in CI/local environments for quicker iterations.

marseel reviewed (changes requested):
Looks good overall, I did mostly high-level review for now.

squeed commented on pkg/kvstore/etcd.go:
```diff
@@ -580,7 +591,30 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
+				RateBurst:        clientOptions.RateLimit,
+				ParallelRequests: clientOptions.MaxInflight,
+			}, ciliumratemetrics.APILimiterObserver())
+			ec.limiter.Swap(newLimiter)
```
super minor nit: use `.Store()` :-)

giorio94 commented on pkg/kvstore/etcd.go:
Nit. The message of commit `etcd: allow threadsafe replacement of limiter` does no longer reflect the actual changes.

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -580,7 +591,30 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
+	if opts != nil && opts.BootstrapComplete != nil {
+		initialLimit = clientOptions.BootstrapRateLimit
+		go func() {
+			<-opts.BootstrapComplete
```
I'd suggest additionally selecting on etcd client context expiration as well, to avoid unnecessarily blocking if the client gets closed. Admittedly not something that should ever happen (as in that case we would be stopping anyways), but a little more future proof anyways.

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -580,7 +591,30 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
+				RateBurst:        clientOptions.RateLimit,
+				ParallelRequests: clientOptions.MaxInflight,
+			}, ciliumratemetrics.APILimiterObserver())
+			ec.limiter.Swap(newLimiter)
```
I wonder whether it wouldn't be better to just expose an APILimiter method to perform such update, instead of creating a brand new rate limiter. That would both simplify the changes here and allow to preserve enforcing the other properties, such as max inflights, also during the transition.

thorn3r replied:
yes good point, let's do that

giorio94 commented on clustermesh-apiserver/health/health.go:
```diff
@@ -10,47 +10,45 @@ import (
+const (
+	defaultHealthPort = 9880
+	optHealthPort     = "health-port"
 )
```
Nit. These constants are only used once, hence they could be inlined as well.

giorio94 commented on clustermesh-apiserver/health/health.go:
```diff
@@ -10,47 +10,45 @@ import (
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "clustermesh-health")
```
Nit. The logger should be [retrieved from hive](https://github.com/cilium/cilium/blob/d72e63048361bcbe85be260c390fdecbc4ad9ff7/pkg/metrics/registry.go#L46), rather than defined as a global variable.

giorio94 commented on clustermesh-apiserver/health/health.go:
```diff
@@ -10,47 +10,45 @@ import (
+func registerHealthAPIServer(lc hive.Lifecycle, cfg HealthAPIServerConfig, syncState *common.SyncState) {
 	mux := http.NewServeMux()
 
-	mux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
```
Are we confident that we don't need these at all? Otherwise it could make sense to start thinking how to make this logic composable, to make easy to specify different health checks for clustermesh and kvstoremesh.

thorn3r replied:
As of now, the /healthz endpoint is unused. It's possible we may want to implement it, but they may also share the same logic (i.e. can't write to the local etcd). I think making it modular would be good future-proofing, though :+1:

giorio94 replied:
Yeah, I was mostly thinking about k8s client health checks which (as of now, although it may change in the future) would not be meaningful in the kvstoremesh context. But I agree that it is something not applicable right now, and we can always postpone generalization to later.

giorio94 commented on pkg/clustermesh/common/sync.go:
```diff
@@ -0,0 +1,99 @@
+type SyncState struct {
+	lock.StoppableWaitGroup
+	Init          spanstat.SpanStat
+	CompletedChan chan struct{}
```
Looks redundant to me. Could just use `swg.WaitChannel()`.

thorn3r replied:
yes, you're right :sweat_smile:

giorio94 commented on pkg/kvstore/backend.go:
```diff
@@ -40,6 +40,10 @@ type ExtraOptions struct {
+
+	// BootstrapComplete is a channel that is closed when the cluster has
+	// finished bootstrapping
+	BootstrapComplete chan struct{}
```
`s/chan struct{}/<-chan struct{}/`

giorio94 commented on pkg/clustermesh/common/remote_cluster.go:
```diff
@@ -132,6 +136,19 @@ func (rc *remoteCluster) releaseOldConnection() {
+	// We execute a goroutine to implement a timeout on connection.
+	// If the timeout has been reached and a connection hasn't been established, we
+	// remove the cluster from the SyncState to prevent it from blocking readiness forever.
+	go func() {
```
This will not work as expected if `restartRemoteConnection` is executed again for whatever reason (e.g., triggered by the watchdog), as there would be leaked goroutines potentially decreasing the counter more than desired on timeout. Similarly, it does not account for the possibility that one of the clusters is removed before connection establishment, as it would still wait for timeout expiration.

giorio94 reviewed (changes requested):
Really nice improvement overall!

I did a first pass and left a few comments inline. Generally speaking, I agree with Casey, the first commit is on the large side, and it mixes both refactoring and new functionalities together, making it difficult to review it properly (and discern which changes should be no-op). What about splitting it in multiple chunks? For instance one with the preparatory changes about the health cell, one with the SyncState stuff, and so on? 

I personally also find the SyncState logic to be pretty complex and difficult to reason about. Especially in terms of possible unintended consequences of operations such as configuring new clusters or removing existing ones while synchronization is in progress. They are admittedly rare corner cases, but the current approach could easily lead to incorrect results. OTOH, I don't have yet any really concrete improvement proposal there, but I'll keep reasoning about it.

thorn3r committed:
Refactor remote cluster sync logic for reuse

The logic for tracking synchronization of resources from remote clusters
is also useful for KVStoreMesh. This moves some of the functionality of
'synced' to a new package, pkg/clustermesh/wait, for reuse with a new
SyncedCommon type.

Co-authored-by: Marco Iorio <marco.iorio@isovalent.com>
Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
add 'synced' for kvstoremesh

Add a 'synced' field to remoteCluster in kvstoremesh to track its
synchronization to the local kvstore. This will be used in a following
commit to add a ReadinessProbe to the kvstoremesh container.

A timeout on remote cluster connection is implemented to prevent
blocking readiness forever if we are unable to establish a connection to
a remote cluster. By default this is set to 15s and can be configured via
the option "per-cluster-ready-timeout".

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
Modularize clustermesh-apiserver's health server

Clustermesh-apiserver's Health API Server has been moved to a new
'health' package and modularized for reuse with KVStoreMesh.

A slice of EndpointFunc must be provided to the HealthAPIServerCell to
be registered as endpoints the Health API will serve.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
kvstoremesh: add readinessProbe on initial sync

Register a '/readyz' endpoint for KVStoreMesh that exposes the status of
the initial synchronization of remote clusters to the local kvstore.
The endpoint will return a 500 until the synchronization has completed,
at which point it will respond with 200.

A new cell added to KVStoreMesh executes a OneShot job to wait for
remote cluster synchronization and signal completion to the
HealthAPIServer via the shared SyncState. There are 2 configurable
timeouts to prevent an unreachable cluster or synchronization error
from blocking readiness forever and potentially causing a cascading
failure as more KVStoreMesh instances restart:

- 'per-cluster-ready-timeout' is the duration to wait for a remote cluster
connection before removing it from the readiness check.

- 'global-ready-timeout' is the duration to wait for all remote clusters
  to synchronize before cancelling the wait and reporting ready.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
APILimiter: Add setters for rate limit and burst

Add 2 new methods to APILimiter that allow the RateLimit and BurstLimit
to be adjusted externally.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
kvstore/etcd: add kvstore-opt etcd.bootstrapQps

Add a new kvstore-opt 'etcd.bootstrapQps' to configure an initial client
rate limit to use until the caller signals that bootstrap is complete.

This is used in ClusterMesh and KVStoreMesh to allow setting a higher
rate limit during the initial synchronization to etcd. To prevent
changing behavior for other components using etcd clients, this option
will only affect components providing a channel via
extraOptions.BootstrapComplete, which should be closed to signal to the
client that bootstrap is complete.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

youngnick reviewed (approved):
Can't find why tophat was added on this one, so I'll get it out of your way.

giorio94 commented on clustermesh-apiserver/option/config.go:
```diff
@@ -52,3 +52,15 @@ var DefaultLegacyKVStoreMeshConfig = LegacyKVStoreMeshConfig{
+}
+
+var DefaultKVStoreMeshConfig = KVStoreMeshConfig{
+	RemoteClusterReadyTimeout: 1 * time.Minute,
```
I'd be personally in favor of lowering the default timeout here to maybe 10-15s, as it only accounts for the initial connection establishment (which should be always fast), and not full data synchronization. Otherwise, we'll introduce unnecessary delays e.g., during the initial deployment, when the connection to the remote cluster would never succeed as it is also initializing at the same time.

giorio94 commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -65,17 +65,19 @@ type ClusterMesh struct {
+	mutex              lock.RWMutex
+	clusters           map[string]*remoteCluster
+	configWatcher      *configDirectoryWatcher
+	InitialConfigsRead chan struct{}
```
`InitialConfigsRead` doesn't look necessary to me, because all configurations are read (and the remote cluster struct created) synchronously by a Start hook, and all start hooks registered by other cells depending on KVStoreMesh are guaranteed to be executed after its termination.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -79,18 +97,24 @@ func (km *KVStoreMesh) Stop(cell.HookContext) error {
+		readyTimeout: km.remoteClusterReadyTimeout,
+		logger:       km.logger.WithField(logfields.ClusterName, name),
 	}
+	synced.Stop()
```
This doesn't look correct. I guess that you'd want to stop the `resources` swg. `Stop()`, instead, should be called when the remote cluster is stopped. Edit, I see it was fixed in a later commit, should be moved to the second one.

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -115,13 +136,34 @@ func (rc *remoteCluster) Remove() {
+	case <-ctx.Done():
+	case <-rc.synced.connected:
+	case <-time.After(rc.readyTimeout):
+		rc.logger.Warning("Remote cluster did not connect within timeout, removing from readiness checks")
```
I'd lower the severity of this log message to info, as there's nothing the user can do about it.

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -133,14 +175,15 @@ func (o *syncer) OnDelete(key store.NamedKey) {
 }
 
-func newReflector(local kvstore.BackendOperations, cluster, prefix string, factory store.Factory) reflector {
+func newReflector(local kvstore.BackendOperations, cluster, prefix string, factory store.Factory, synced *lock.StoppableWaitGroup) reflector {
```
You should call `synced.Add()` somewhere in this function. Edit. It appears to be in the fifth commit, should be moved to the second one.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh_test.go:
```diff
@@ -197,7 +198,7 @@ func TestRemoteClusterRun(t *testing.T) {
 			}
 			st := store.NewFactory(store.MetricsProvider())
-			km := KVStoreMesh{backend: kvstore.Client(), storeFactory: st}
+			km := KVStoreMesh{backend: kvstore.Client(), storeFactory: st, logger: logrus.New()}
```
It would be nice to test at least the happy case.

thorn3r replied:
Should have good coverage now. I added an interface for common.ClusterMesh so that I could mock it and be able to simulate long synchronizations

giorio94 commented on clustermesh-apiserver/syncstate/syncstate.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func MetricsProvider() func() Metrics {
+	return func() Metrics {
```
This extra function looks unnecessary. You could simply return the Metrics object.

giorio94 commented on clustermesh-apiserver/syncstate/syncstate.go:
```diff
@@ -0,0 +1,77 @@
+	InitialSyncDuration metric.Gauge
+}
+
+func MetricsProvider() func() Metrics {
```
You should reflect the new metric in `Documentation/observability/metrics.rst`.

giorio94 commented on clustermesh-apiserver/clustermesh/health.go:
```diff
@@ -4,81 +4,62 @@
+				}
+				w.WriteHeader(statusCode)
+				if _, err := w.Write([]byte(reply)); err != nil {
+					log.WithError(err).Error("Failed to respond to /readyz request")
```
Nit. Could use the logger retrieved from hive instead of the global one.

giorio94 commented on clustermesh-apiserver/syncstate/syncstate_test.go:
```diff
@@ -0,0 +1,41 @@
+		assert.True(c, ss.Complete())
+		_, ok := <-ss.WaitChannel()
+		assert.False(c, ok)
+	}, timeout, tick, "Expected SyncState to be completed")
```
Could be tested synchronously (without `EnvetuallyWithTf`), as it doesn't depend on any background work. The two assertions look also redundant to me, as testing the same aspect.

giorio94 commented on pkg/rate/api_limiter_test.go:
```diff
@@ -794,3 +794,24 @@ func (b *ControllerSuite) TestReservationCancel(c *check.C) {
 	}
 }
+
+func (b *ControllerSuite) TestSetRateLimit(c *check.C) {
```
New tests should use the testing framework instead of the checkmate, unless strictly necessary.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -98,6 +129,7 @@ func (km *KVStoreMesh) newRemoteCluster(name string, _ common.StatusFunc) common
 	ctx, cancel := context.WithCancel(context.Background())
 
 	synced := newSynced(ctx)
+	defer synced.resources.Stop()
```
Should be moved to the second commit.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -135,7 +166,8 @@ func (km *KVStoreMesh) newRemoteCluster(name string, _ common.StatusFunc) common
 // synced returns after the initial cluster configs have been read and all
-// remote clusters have been synchronized.
+// remote clusters have been synchronized. The given syncCallback is always
+// executed before the function returns.
```
Should be moved to the second commit.

giorio94 commented on clustermesh-apiserver/option/config.go:
```diff
@@ -54,13 +54,16 @@ func (def LegacyKVStoreMeshConfig) Flags(flags *pflag.FlagSet) {
 var DefaultKVStoreMeshConfig = KVStoreMeshConfig{
-	RemoteClusterReadyTimeout: 1 * time.Minute,
+	PerClusterReadyTimeout: 1 * time.Minute,
+	GlobalReadyTimeout:     2 * time.Minute,
```
I'd personally raise this default to an higher value to account for large scale environments as well. Maybe something in the 5-10m range?

giorio94 commented on clustermesh-apiserver/option/config.go:
```diff
@@ -54,13 +54,16 @@ func (def LegacyKVStoreMeshConfig) Flags(flags *pflag.FlagSet) {
 
 func (def KVStoreMeshConfig) Flags(flags *pflag.FlagSet) {
-	flags.Duration("remote-cluster-ready-timeout", def.RemoteClusterReadyTimeout, "Remote clusters will be disregarded for readiness checks if a connection cannot be established within this duration")
+	flags.Duration("per-cluster-ready-timeout", def.PerClusterReadyTimeout, "Remote clusters will be disregarded for readiness checks if a connection cannot be established within this duration")
```
The first parameter should be changed in the second commit, instead of here.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -39,6 +41,8 @@ type KVStoreMesh struct {
 	// established.
 	remoteClusterReadyTimeout time.Duration
 
+	globalReadyTimeout time.Duration
```
Both this field and the one above are duplicates of `option.KVStoreMeshConfig`.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -80,6 +86,31 @@ func newKVStoreMesh(lc cell.Lifecycle, params params) *KVStoreMesh {
+
+func RegisterSyncWaiter(lc cell.Lifecycle, p SyncWaiterParams) {
+	lc.Append(cell.Hook{
+		OnStart: func(ctx cell.HookContext) error {
```
Blocking a start hook will prevent every other start hook from being executed. Additionally, there's a timeout causing the shutdown of the entire hive if they don't terminate on time (IIRC something around 5 minutes).

thorn3r replied:
I unblocked the start hook by spawning a new context. This is probably not ideal, but using a start hook is pretty convenient for making sure it runs after the common.ClusterMesh hook completes and the configs have been read.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -80,6 +86,31 @@ func newKVStoreMesh(lc cell.Lifecycle, params params) *KVStoreMesh {
+func RegisterSyncWaiter(lc cell.Lifecycle, p SyncWaiterParams) {
+	lc.Append(cell.Hook{
+		OnStart: func(ctx cell.HookContext) error {
+			syncCtx, cancel := context.WithTimeout(ctx, p.KVStoreMesh.config.GlobalReadyTimeout)
```
I'd move the timeout to the synced function, given that the parameter is part of kvstoremesh nonetheless.

giorio94 reviewed (changes requested):
I did a first pass, and overall looks way better to me. I've left a few comments inline, mostly nits (especially around some changes having been mixed up between the second and the fifth commit).

thorn3r committed:
refactor pkg/rate tests to use testify framework

The following commit will add new functionality to APILimiter and
coverage with tests using testify. To avoid dropping in new tests that
don't match the rest of the suite, all of the pkg/rate tests have been
refactored to use the testify framework.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
refactor common.ClusterMesh to an interface

Make common.ClusterMesh an interface so that it can be mocked in
integration tests. The original type is now an exported struct,
common.clusterMesh.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -1115,6 +1115,7 @@ Name                                     Labels
 ``kvstore_sync_errors_total``            ``scope``, ``source_cluster``                Number of times synchronization to the kvstore failed
 ``kvstore_sync_queue_size``              ``scope``, ``source_cluster``                Number of elements queued for synchronization in the kvstore
 ``kvstore_initial_sync_completed``       ``scope``, ``source_cluster``, ``action``    Whether the initial synchronization from/to the kvstore has completed
+``kvstore_initial_sync_time_seconds``                                                 Duration in seconds of the initial sync to the kvstore for local cluster resources
```
Just double-checking to be sure: No related labels for these new metrics?

thorn3r replied:
Good question! Perhaps an `outcome` would be useful here to record whether we hit either timeout. Now I am wondering though if this belongs with this group of metrics at all. If the sync passes either timeout, it will look strange plotted against `kvstore_initial_sync_completed`. This one tracks the synchronization of all remote clusters to the local kvstore, while the other sync-related ones track an individual remote cluster. The metric is really more of an indication of how long it took to be ready to serve requests.

Maybe this should be moved out of the kvstore grouping and renamed to `bootstrap_seconds`?

thorn3r replied:
Renamed to `bootstrap_seconds` and added a `source_cluster` label. Decided against adding `outcome` to surface timeouts. It added complexity and that info is already available  via the other metrics.

giorio94 replied:
This entry should be removed then, I guess.

giorio94 commented on clustermesh-apiserver/option/config.go:
```diff
@@ -52,3 +52,15 @@ var DefaultLegacyKVStoreMeshConfig = LegacyKVStoreMeshConfig{
+}
+
+var DefaultKVStoreMeshConfig = KVStoreMeshConfig{
+	PerClusterReadyTimeout: 15 * time.Second,
```
Nit. The commit message should be amended as well, as still refers to 1m:

> By default this is set to 1m and can be configured via
the option "remote-cluster-ready-timeout".

giorio94 commented on clustermesh-apiserver/syncstate/syncstate_test.go:
```diff
@@ -0,0 +1,31 @@
+	ss.Stop()
+
+	// call all doneFuncs to simulate the resources being synchronized
+	for _, f := range doneFuncs {
```
Could add `require.False(t, ss.Complete())` to assert that `Complete()` correctly returns false as well.

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -1166,6 +1177,16 @@ Exported Metrics
+``bootstrap_seconds``                    ``source_cluster``                           Duration in seconds to complete bootstrap
+======================================== ============================================ ========================================================
+
+
```
Should be moved to the correct commit.

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -1211,6 +1211,7 @@ Name                                     Labels
 ``kvstore_sync_errors_total``            ``scope``, ``source_cluster``                Number of times synchronization to the kvstore failed
 ``kvstore_sync_queue_size``              ``scope``, ``source_cluster``                Number of elements queued for synchronization in the kvstore
 ``kvstore_initial_sync_completed``       ``scope``, ``source_cluster``, ``action``    Whether the initial synchronization from/to the kvstore has completed
+``kvstore_initial_sync_time_seconds``                                                 Duration in seconds of the initial sync to the kvstore for remote cluster resources
```
Should be dropped.

giorio94 commented on clustermesh-apiserver/option/config.go:
```diff
@@ -55,12 +55,15 @@ func (def LegacyKVStoreMeshConfig) Flags(flags *pflag.FlagSet) {
 
 func (def KVStoreMeshConfig) Flags(flags *pflag.FlagSet) {
 	flags.Duration("per-cluster-ready-timeout", def.PerClusterReadyTimeout, "Remote clusters will be disregarded for readiness checks if a connection cannot be established within this duration")
+	flags.Duration("global-ready-timeout", def.GlobalReadyTimeout, "KVStoreMesh will be considered ready if any remote clusters have failed to synchronize within this duration")
```
Nit. `s/considered ready if any/considered ready even if any/`

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -74,6 +80,34 @@ func newKVStoreMesh(lc cell.Lifecycle, params params) *KVStoreMesh {
+				p.SyncState.Stop()
+			}
+
+			go p.KVStoreMesh.synced(ctx, syncedCallback)
```
Nit. You should cancel the context in a stop hook, and wait for the termination of the go routine. Or alternatively use a `OneShot` job.

thorn3r replied:
I changed it to a OneShot job, which seems more suitable for this. Thanks for the suggestion!

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -128,18 +162,37 @@ func (km *KVStoreMesh) newRemoteCluster(name string, _ common.StatusFunc) common
-		waiters = append(waiters, rc.synced.Resources)
-		return nil
-	})
+	ctx, cancel := context.WithCancel(ctx)
```
Unless I'm missing something, you could simply use `context.WithTimeout()` and drop the rest of the changes.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh_test.go:
```diff
@@ -251,3 +256,174 @@ func TestRemoteClusterRun(t *testing.T) {
+
+			wg.Add(1)
+			go func() {
+				time.Sleep(tt.connectDelay)
```
I'm personally opposed to unit tests which depend too heavily on time, as typically flaky in CI environments. I'd also avoid unit tests that require multiple seconds to complete, as they unnecessarily extend the total duration of workflows, and make troubleshooting possible flakes really difficult. 

Is it really necessary relying on real time here? A common pattern I've seen in the past consists in assigning `time.After` to a variable, and replacing that in tests with a mocked version of it which plays with channels for synchronization.

```
--- PASS: TestRemoteClusterSync (15.12s)
    --- PASS: TestRemoteClusterSync/remote_cluster_connection_success (0.02s)
    --- PASS: TestRemoteClusterSync/remote_cluster_connection_timeout (10.04s)
    --- PASS: TestRemoteClusterSync/remote_cluster_sync_timeout (5.04s)
```

thorn3r replied:
I redid the entire test to be unit tests. I initially used your suggestion of replacing the timeouts with channels, but it ended up being a bit simpler to just dial the timeouts down to almost 0 and manipulate the `synced` and `syncstate` to produce the desired behavior. Let me know what you think about this approach!
```
=== RUN   TestRemoteClusterSync
=== RUN   TestRemoteClusterSync/remote_cluster_successfully_syncs
=== RUN   TestRemoteClusterSync/remote_cluster_fails_to_connect
time="2024-02-28T23:58:29-05:00" level=info msg="Remote cluster did not connect within timeout, removing from readiness checks" clusterName=foo
time="2024-02-28T23:58:29-05:00" level=info msg="Some clusters may not have been synchronized" error="context deadline exceeded"
=== RUN   TestRemoteClusterSync/remote_cluster_connects_but_fails_to_sync
time="2024-02-28T23:58:29-05:00" level=info msg="Some clusters may not have been synchronized" error="context deadline exceeded"
--- PASS: TestRemoteClusterSync (0.00s)
    --- PASS: TestRemoteClusterSync/remote_cluster_successfully_syncs (0.00s)
    --- PASS: TestRemoteClusterSync/remote_cluster_fails_to_connect (0.00s)
    --- PASS: TestRemoteClusterSync/remote_cluster_connects_but_fails_to_sync (0.00s)
PASS
ok      github.com/cilium/cilium/pkg/clustermesh/kvstoremesh    0.026s
```

giorio94 commented on pkg/rate/limiter_test.go:
```diff
@@ -8,50 +8,28 @@ import (
-	c.Assert(l.AllowN(100), Equals, true)
-	// We shouldn't be allowed to get any left
-	c.Assert(l.Allow(), Equals, false)
+	assert.True(t, l.AllowN(100), "Limiter should have 100 available tokens")
```
`c.Assert()` and the `assert` package are not equivalent, because the latter generates non-blocking errors (i.e., the rest of the test continues to be executed, and errors are displayed at the end). You may want to use the `require` package to mimic the same behavior (all the methods are the same).

IMHO, which one to use is mostly a matter of preference. One reason I typically use `require` is that in many cases if an error occurs at a certain point then the subsequent operations are also likely to fail, and it makes it easier to determine the source cause. `assert`, instead, should be preferred for independent assertions.

thorn3r replied:
Good call out on this one :+1:

giorio94 commented on pkg/rate/api_limiter_test.go:
```diff
@@ -130,60 +137,63 @@ func (b *ControllerSuite) TestMeanProcessingDuration(c *check.C) {
 
-	c.Assert(a.requestsProcessed, check.Equals, int64(10))
-	c.Assert(a.parallelRequests, check.Equals, 2)
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
```
Does this need to be run in an `Eventually` block? It wasn't before.

giorio94 commented on pkg/rate/api_limiter_test.go:
```diff
@@ -470,149 +477,160 @@ func (b *ControllerSuite) TestAPILimiterMetrics(c *check.C) {
 }
 
-func (b *ControllerSuite) TestAPILimiterMergeUserConfig(c *check.C) {
+// FIX THIS
```
I guess this is a leftover comment.

giorio94 commented on pkg/rate/api_limiter_test.go:
Thanks for taking care of this!

giorio94 reviewed (commented):
A bunch of minor comments inline, before starting the final testing.

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -580,6 +590,29 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
+		initialLimit = clientOptions.BootstrapRateLimit
+		go func() {
+			select {
+			case <-ctx.Done():
```
I'd replace this with `ec.client.Ctx().Done()`, as ctx is not guaranteed to be closed if the `client.Close()` gets called.

giorio94 commented on clustermesh-apiserver/option/config.go:
```diff
@@ -52,3 +52,18 @@ var DefaultLegacyKVStoreMeshConfig = LegacyKVStoreMeshConfig{
+type KVStoreMeshConfig struct {
+	PerClusterReadyTimeout time.Duration
+	GlobalReadyTimeout     time.Duration
+}
```
I didn't notice this earlier, but configurations are typically defined in the same package which registers and uses them. So in this case it would be `pkg/clustermesh/kvstoremesh`.

giorio94 reviewed (changes requested):
Requesting changes so that this disappears form my review queue. Feel free to re-request my review once ready again.

giorio94 commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -60,8 +60,20 @@ type Configuration struct {
+	NumReadyClusters() int
+
+	Start(cell.HookContext) error
+	Stop(cell.HookContext) error
```
Nit. Could be replaced by `cell.HookInterface`

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -88,6 +94,30 @@ func newKVStoreMesh(lc cell.Lifecycle, params params) *KVStoreMesh {
+	jobGroup.Add(
+		job.OneShot("kvstoremesh-sync-waiter", func(ctx context.Context, health cell.HealthReporter) error {
+			syncedCallback := p.SyncState.WaitForResource()
+			p.SyncState.Stop()
```
There's a (admittedly small) potential for a race condition here. Jobs are [started asynchronously](https://github.com/thorn3r/cilium/blob/ec21e032f3bad8dd660cca6e3e4e591cc2ef5093/pkg/hive/job/job.go#L162), and the sync state may be checked before the wait group count is increased. Moving these two lines to the main `RegisterSyncWaiter` funciton body would fix it.

thorn3r replied:
I think this should be fine, since the syncstate cannot be "complete" until `Stop()` is called on the underlying swg, but I moved it into the main function anyways

giorio94 replied:
Yeah, you are right. :+1:

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -141,3 +171,28 @@ func (km *KVStoreMesh) newRemoteCluster(name string, _ common.StatusFunc) common
+	})
+
+	if err := wait.ForAll(ctx, waiters); err != nil {
+		km.logger.WithError(err).Info("Some clusters may not have been synchronized")
```
Nit. Could be worth extending the message to state the consequence (i.e, kvstoremesh is marked ready regardless).

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh_test.go:
```diff
@@ -251,3 +256,114 @@ func TestRemoteClusterRun(t *testing.T) {
+			}
+			km := KVStoreMesh{
+				// Set a very low timeout to speed up the test
+				config: Config{PerClusterReadyTimeout: 1 * time.Millisecond, GlobalReadyTimeout: 1 * time.Millisecond},
```
I'd suggest using different timeout values depending on whether they are expected to be hit or not. That way, you speed up the test for expected failures, but also make sure that timeout is not hit by mistake (hence causing a flake) in the happy case.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh_test.go:
```diff
@@ -251,3 +256,114 @@ func TestRemoteClusterRun(t *testing.T) {
+			rc := &remoteCluster{
+				name:   "foo",
+				synced: synced,
+				logger: km.logger.WithField(logfields.ClusterName, "foo"),
```
The `PerClusterReadyTimeout` should be configured here, rather than in `km`.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh_test.go:
```diff
@@ -251,3 +256,114 @@ func TestRemoteClusterRun(t *testing.T) {
+			// trigger the readiness timeout
+			rc.waitForConnection(ctx)
+
+			require.NoError(t, rc.synced.Resources(ctx), "Still waiting for remote cluster resources")
```
This check passes even if the `if tt.connect` branch above is removed. It could make sense to amend the test so that it is more robust. One possibility could be avoiding to call `synced.resources.Done()` above and here assert that the wait group got decremented only in the `!connected` case.

thorn3r replied:
This is intentional, since all resources should be completed by this point in both paths:
- The happy path is that the connection succeeds, making `rc.waitForConnection` a noop and returning immediately.
- The sad path is we don't establish a connection, and `rc.WaitForConnection` triggers the timeout, marking all resources as done.

There may be another angle I'm not seeing, though

thorn3r replied:
After re-reading your comment I realize I misunderstood. If the `tt.connect` branch is removed, cases where the conditional is true will still pass, but due to the timeout rather than actually being marked `Done()`. I'll update the test

thorn3r replied:
Updated. If the `tt.connect` conditional is removed, tests that expect `tt.connect` will fail

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2560,8 +2560,19 @@ clustermesh:
+      # -- duration to wait before the first probe
+      initialDelaySeconds: 5
+      # -- timeout for the readiness probe
+      timeoutSeconds: 5
```
Any reason for not using kubernetes default values here? (Same for kvstoremesh). I'd also configure at least `contrib/testing/kind-clustermesh{1,2}.yaml` with more aggressive values (e.g. `1 second period`).

thorn3r replied:
The default for `periodsSeconds` is 10, which seems too aggressive for a check that we really only need to pass once but will continue to run during the lifetime of the Pod. It would be more convenient if there was something similar to a startupProbe but for readiness. Default for initial delay is 0, which would likely cause the pod to take at least 30 seconds to become ready in most cases.

The other 2 are the same as the default setting, but it doesn't seem like a bad thing to have those values configurable with the rest.

giorio94 replied:
> The default for periodsSeconds is 10, which seems too aggressive for a check that we really only need to pass once but will continue to run during the lifetime of the Pod.

I'd personally argue that this is the case with quite a few readiness probes (and regardless of whether it is expected to change or not, the amount of requests is the same), and a local HTTP request every 10 seconds should be more than fine, especially given that the handler is practically a no-op in terms of overhead. But that's mostly a matter of opinions I guess.

> It would be more convenient if there was something similar to a startupProbe but for readiness.

Totally agree. It should be pretty easy to configure a `startupProbe` for kvstoremesh, given the global circuit breaker timeout, but it is way more trickier for the clustermesh-apiserver. And overall the risks associated with the startupProbe terminating the containers overweight the advantages.

> Default for initial delay is 0, which would likely cause the pod to take at least 30 seconds to become ready in most cases.

The initial delay doesn't appear to work as I (and based on the above also you) expected. Probing is performed according to the usual schedule, but [the result is ignored if it falls during the initial delay window](https://github.com/kubernetes/kubernetes/blob/4b8e819355d791d96b7e9d9efe4cbafae2311c88/pkg/kubelet/prober/worker.go#L268-L271). Hence, the pod will effectively always require more than 30 seconds to become ready. Which is also confirmed by local testing.

> The other 2 are the same as the default setting, but it doesn't seem like a bad thing to have those values configurable with the rest.

Sure, makes totally sense to me.

squeed replied:
can the kvstoremesh ever transition from Ready to NotReady? If not, a StartupProbe may be the correct answer.

giorio94 replied:
> can the kvstoremesh ever transition from Ready to NotReady? 

Currently not (same for the clustermesh-apiserver). And likely not even in the near future, while we may have a readiness probe for the etcd possibly transitioning into NotReady state.

>  If not, a StartupProbe may be the correct answer.

I agree, but the main issue with the startup probe is that it behaves like a liveness probe, restarting the container in case of failure, which would cause a lot of issues in this case -- as the synchronization would need to be performed again from scratch. The only possibility I see could be setting an extremely high failureThreshold, as depending on the number of QPS and cluster size the initial synchronization may take in the order of several (tens of) minutes -- although the bootstrap QPS logic implemented here helps a lot -- which is suboptimal either.

thorn3r replied:
Given that info, I've updated the chart to use all default values for the probe unless provided via helm values

giorio94 reviewed (changes requested):
Overall looks good to me, thanks. Please find my last round of feedback inline, then it should be good to go from my point of view.

marseel commented on clustermesh-apiserver/clustermesh/health.go:
```diff
@@ -4,81 +4,65 @@
+			},
+		},
+		{
+			Path: "/healthz",
```
I would consider removing endpoint `ServerVersion` at this point as it seems it's doing exactly the opposite of what it should - I know that is the way it was implemented before, but I can't see any reasonable use case for checking k8s-apiserver health through clustermesh-apiserver.  If someone would use it for health-checking then it would be even worse, when k8s control-plane goes down - all clustermesh-apiservers go down as well and new agents in other clusters can't fetch any data.

tldr until we figure out better health checking I would just remove `ServerVersion` check here.

thorn3r replied:
I initially removed it, then there was a suggestion to not remove it so I replaced it. I will remove it again :rofl:

marseel commented on clustermesh-apiserver/syncstate/syncstate.go:
```diff
@@ -0,0 +1,76 @@
+
+func new(metrics Metrics, clusterInfo types.ClusterInfo) SyncState {
+	ss := SyncState{StoppableWaitGroup: lock.NewStoppableWaitGroup()}
+	syncTime := spanstat.Start()
```
nit: we could move it inside `go func`

marseel reviewed (commented):
I will take one more in-depth look on Monday, but overall looks good to me :+1:

thorn3r committed:
remove unused clustermesh /healthz endpoint

The /healthz endpoint for ClusterMesh is not currently used. If someone
were to configure a livenessProbe targeting this endpoint, it would
cause clustermesh-apiserver to fail if the kube-apiserver was unhealthy.
This would prevent new agents in other clusters from being able to
synchronize data.

For now, we will remove this endpoint to prevent this scenario.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

giorio94 commented on contrib/testing/kind-clustermesh1.yaml:
```diff
@@ -29,3 +29,12 @@ clustermesh:
+    readinessProbe:
+      periodSeconds: 1
+      initialDelaySeconds: 1
+      failureThreshold: 30
```
Nit. This is in the wrong commit (same for kind-clustermesh2), as the clustermesh probe is made configurable only in a later commit.

giorio94 commented on contrib/testing/kind-clustermesh1.yaml:
```diff
@@ -29,3 +29,12 @@ clustermesh:
+    readinessProbe:
+      periodSeconds: 1
+      initialDelaySeconds: 1
+      failureThreshold: 30
```
Nit. Is it necessary to configure a high failure threshold?

thorn3r replied:
Nope, was thinking about it like a livenessProbe. Fixed

marseel commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -79,7 +80,7 @@ type parameters struct {
 	Resources      cmk8s.Resources
 	BackendPromise promise.Promise[kvstore.BackendOperations]
 	StoreFactory   store.Factory
-	SyncState      *SyncState
```
I think you forgot to remove `NewSyncState` below or maybe I am missing something?
Also, shouldn't it stay as a pointer? It seems to me that it works right now only because of how `SyncState` is implemented (i. e. only pointer `*lock.StoppableWaitGroup`) and essentially it's kind of a singleton.

thorn3r replied:
Good call, I missed this while redoing this since it wasn't in the diffs :sweat_smile: 

I changed it from SyncState needing to be used as a pointer, to just making the internal SWG a pointer. I thought this would be a bit safer, guaranteeing that the SyncState will always use the same SWG whether pass-by-value or pointer. You could still create another SyncState with a different SWG by using `syncstate.Cell`

marseel commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -105,5 +167,32 @@ func (km *KVStoreMesh) newRemoteCluster(name string, _ common.StatusFunc) common
+	})
+
+	if err := wait.ForAll(ctx, waiters); err != nil {
+		km.logger.WithError(err).Info("Failed to wait for synchronization. KVStoreMesh will now handle requests, but some clusters may not have been synchronized.")
```
nit: Probably could be a warning and also it would be nice to provide information for which cluster it happened for easier debugging (Doesn't need to be a list of all clusters, just an example will be helpful I think)

thorn3r replied:
Made this Info level because it seemed similar in vein to the connection timeout, since it doesn't necessarily mean the synchronization itself failed, just that it is taking longer than expected: https://github.com/cilium/cilium/pull/30361#discussion_r1489590058

We could add this info to the log, but the data is also available via the `kvstore_initial_sync_completed` metric that kvstoremesh exports

marseel reviewed (approved):
Thanks, looks good from my side. Two last comments, but I am approving as they are non-blocking.

bimmlerd reviewed (approved):
contributing changes reviewed by Marco already - rubber stamping as a means of deferring to his review

squeed reviewed (approved):
one minor question, but my files look good.

thorn3r committed:
helm: add readinessProbe to kvstoremesh

Add a readinessProbe to the kvstoremesh container in the
clustermesh-apiserver deployment utilizing the '/readyz' endpoint of
the health API.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
helm: make ClusterMesh readinessProbe configurable

Similar to KVStoreMesh, the settings for clustermesh-apiserver's
readinessProbe should be configurable via helm values.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2562,8 +2562,13 @@ clustermesh:
+
     # -- TCP port for the clustermesh-apiserver health API.
     healthPort: 9880
+    # readinessProbe:
```
Nit. I wonder if it could make sense to uncomment `readinessProbe:` (adding `{}` at the end), so that it is possible to add a comment, and it appears as part of the readme. Same for kvstoremesh.

thorn3r replied:
updated :+1:

giorio94 reviewed (approved):
Thanks, lgtm!

joestringer reviewed (approved):
Providing the required 0 reviews to satisfy our GitHub overlords.

(We may want to update CODEOWNERS for `pkg/rate` if we expect further changes there, would skip tophat requirement in future)
